diff -r -N -w -u -I '[$]Id:' --exclude='*.class' --exclude='.*' --exclude='*.orig' --exclude='*.rej' --exclude=CVS --exclude='#*' --exclude='*~' ./MMTk/src/org/mmtk/utility/statistics/Stats.java ./MMTk/src/org/mmtk/utility/statistics/Stats.java
--- ./MMTk/src/org/mmtk/utility/statistics/Stats.java	2006-11-26 16:07:47.000000000 +1100
+++ ./MMTk/src/org/mmtk/utility/statistics/Stats.java	2006-11-26 16:11:30.000000000 +1100
@@ -35,7 +35,7 @@
    * Class variables
    */
 
-  public static final boolean GATHER_MARK_CONS_STATS = false;
+  public static final boolean GATHER_MARK_CONS_STATS = true;
 
   /** Maximum number of gc/mutator phases that can be counted */
   static final int MAX_PHASES = 1 << 12;
@@ -90,8 +90,6 @@
         counter[c].phaseChange(phase);
       }
       phase++;
-    } else {
-      Log.writeln("Warning: number of GC phases exceeds MAX_PHASES");
     }
   }
 
@@ -106,8 +104,6 @@
         counter[c].phaseChange(phase);
       }
       phase++;
-    } else {
-      Log.writeln("Warning: number of GC phases exceeds MAX_PHASES");
     }
   }
 
