diff -r -w -u -N -I '[$]Id:' -I '[$]Revision:' -I '[$]Date:' --exclude='*.orig' --exclude=CVS --exclude='*.rej' --exclude='*~' base/rvm/src/vm/compilers/harness/runtime/VM_RuntimeCompiler.java rvm/src/vm/compilers/harness/runtime/VM_RuntimeCompiler.java
--- base/rvm/src/vm/compilers/harness/runtime/VM_RuntimeCompiler.java	2005-07-18 05:57:39.000000000 +1000
+++ rvm/src/vm/compilers/harness/runtime/VM_RuntimeCompiler.java	2006-06-29 11:50:26.000000000 +1000
@@ -5,6 +5,7 @@
 package com.ibm.JikesRVM;
 
 import com.ibm.JikesRVM.classloader.*;
+import com.ibm.JikesRVM.memoryManagers.mmInterface.MM_Interface;
 //-#if RVM_WITH_ADAPTIVE_SYSTEM
 import com.ibm.JikesRVM.opt.*;
 import com.ibm.JikesRVM.adaptive.*;
@@ -316,7 +317,9 @@
       start = VM_Thread.getCurrentThread().accumulateCycles();
     }
 
+    MM_Interface.enterAllocationScope(MM_Interface.COMPILER_SCOPE);
     VM_CompiledMethod cm = VM_BaselineCompiler.compile(method);
+    MM_Interface.leaveAllocationScope();
 
     if (VM.MeasureCompilation || VM.BuildForAdaptiveSystem) {
       long end = VM_Thread.getCurrentThread().accumulateCycles();
@@ -374,7 +377,9 @@
       start = VM_Thread.getCurrentThread().accumulateCycles();
     }
     
+    MM_Interface.enterAllocationScope(MM_Interface.COMPILER_SCOPE);
     VM_CompiledMethod cm = OPT_Compiler.compile(plan);
+    MM_Interface.leaveAllocationScope();
 
     if (VM.MeasureCompilation || VM.BuildForAdaptiveSystem) {
       long end = VM_Thread.getCurrentThread().accumulateCycles();
diff -r -w -u -N -I '[$]Id:' -I '[$]Revision:' -I '[$]Date:' --exclude='*.orig' --exclude=CVS --exclude='*.rej' --exclude='*~' base/rvm/src/vm/memoryManagers/JMTk/vmInterface/MM_Interface.java rvm/src/vm/memoryManagers/JMTk/vmInterface/MM_Interface.java
--- base/rvm/src/vm/memoryManagers/JMTk/vmInterface/MM_Interface.java	2006-06-19 16:10:11.000000000 +1000
+++ rvm/src/vm/memoryManagers/JMTk/vmInterface/MM_Interface.java	2006-06-29 11:50:26.000000000 +1000
@@ -56,6 +56,7 @@
 import com.ibm.JikesRVM.VM_ObjectModel;
 import com.ibm.JikesRVM.VM_Processor;
 import com.ibm.JikesRVM.VM_Scheduler;
+import com.ibm.JikesRVM.VM_Thread;
 
 /**
  * The interface that the JMTk memory manager presents to the Jikes
@@ -73,6 +74,16 @@
    */
 
   /**
+   * Default scope for policy-based allocation
+   */
+  public static final int DEFAULT_SCOPE = SelectedPlan.DEFAULT_SCOPE;
+  
+  /**
+   * Scope for the compiler thread
+   */
+  public static final int COMPILER_SCOPE = SelectedPlan.COMPILER_SCOPE;
+
+  /**
    * <code>true</code> if a write barrier is required.  Specifically,
    * if the memory manger requires that the virtual machine calls
    * putfieldWriteBarrier, arrayStoreWriteBarrier or modifyCheck when
@@ -220,6 +231,36 @@
 
   /***********************************************************************
    *
+  * Dynamically scoped allocation policies
+  */
+
+  /**
+    * Enter a dynamic allocation scope
+    *
+    * @param scope The allocation scope
+    */
+   public static void enterAllocationScope(int scope) {
+     SelectedMutatorContext.get().setAllocationScope(scope);
+     VM_Thread.getCurrentThread().pushAllocationScope(scope);
+   }
+  
+  /**
+   * Leave a dynamic allocation scope
+   */
+  public static void leaveAllocationScope() {
+    SelectedMutatorContext.get().setAllocationScope(VM_Thread.getCurrentThread().popAllocationScope());
+  }
+
+  /**
+   * Thread switch
+   */
+  public static void threadSwitch(VM_Thread newThread) {
+    SelectedMutatorContext.get().setAllocationScope(newThread.getCurrentAllocationScope());
+  }
+
+  
+  /***********************************************************************
+   *
    * Write barriers
    */
 
diff -r -w -u -N -I '[$]Id:' -I '[$]Revision:' -I '[$]Date:' --exclude='*.orig' --exclude=CVS --exclude='*.rej' --exclude='*~' base/rvm/src/vm/scheduler/VM_Processor.java rvm/src/vm/scheduler/VM_Processor.java
--- base/rvm/src/vm/scheduler/VM_Processor.java	2006-06-19 16:10:12.000000000 +1000
+++ rvm/src/vm/scheduler/VM_Processor.java	2006-06-29 11:50:26.000000000 +1000
@@ -208,6 +208,8 @@
     previousThread.endQuantum(now);
     newThread.startQuantum(now);
     
+    MM_Interface.threadSwitch(newThread);
+    
     //-#if RVM_WITH_HPM
     // set start time of thread
     newThread.startOfWallTime = VM_Magic.getTimeBase();
diff -r -w -u -N -I '[$]Id:' -I '[$]Revision:' -I '[$]Date:' --exclude='*.orig' --exclude=CVS --exclude='*.rej' --exclude='*~' base/rvm/src/vm/scheduler/VM_Thread.java rvm/src/vm/scheduler/VM_Thread.java
--- base/rvm/src/vm/scheduler/VM_Thread.java	2006-06-28 20:55:06.000000000 +1000
+++ rvm/src/vm/scheduler/VM_Thread.java	2006-06-29 11:50:26.000000000 +1000
@@ -1218,6 +1218,8 @@
     this.stack = stack;
     this.thread = thread;
     this.myName = myName;
+    allocationScopeStack = new int [MAX_SCOPE_STACK_DEPTH];
+    allocationScopeStack[0] = MM_Interface.DEFAULT_SCOPE;
 
     chosenProcessorId = (VM.runningVM ? VM_Processor.getCurrentProcessorId() : 0); // for load balancing
     suspendLock = new VM_ProcessorLock();
@@ -2020,4 +2022,48 @@
   // the organizer clear the requests
   public boolean requesting_osr = false;
   //-#endif 
+
+
+  /**
+   * Methods supporting scoped allocation policy
+   */
+  private int[] allocationScopeStack;
+  private int scopeStackPointer = 0;
+  private static final int MAX_SCOPE_STACK_DEPTH = 100; // arbitrary
+
+  /**
+   * Push a policy onto the allocation scope stack
+   *
+   * @param policy The policy to be pushed onto the scope stack
+   */
+  public void pushAllocationScope(int scope) {
+    if (scopeStackPointer < MAX_SCOPE_STACK_DEPTH)
+      allocationScopeStack[++scopeStackPointer] = scope;
+    else
+      VM.sysFail("exceeded allocation scope stack depth"); // !!TODO: grow scope stack
+  }
+
+  /**
+   * Pop the top of the allocation scope stack
+   *
+   * @return The current top of the allocation scope stack
+   */
+  public int popAllocationScope() {
+    if (scopeStackPointer > 0)
+      return allocationScopeStack[--scopeStackPointer];
+    else {
+      VM.sysFail("scope stack underflow");
+      return -1;
+    }
+  }
+
+  /**
+   * Return the current scoped allocation policy
+   *
+   * @return The policy at the top of the scoped allocation policy stack
+   */
+  public int getCurrentAllocationScope() {
+    if (VM.VerifyAssertions) VM._assert(scopeStackPointer >= 0);
+    return allocationScopeStack[scopeStackPointer];
+  }
 }
diff -r -w -u -N -I '[$]Id:' -I '[$]Revision:' -I '[$]Date:' --exclude='*.orig' --exclude=CVS --exclude='*.rej' --exclude='*~' base/MMTk/src/org/mmtk/plan/MutatorContext.java MMTk/src/org/mmtk/plan/MutatorContext.java
--- base/MMTk/src/org/mmtk/plan/MutatorContext.java	2006-06-21 17:45:17.000000000 +1000
+++ MMTk/src/org/mmtk/plan/MutatorContext.java	2006-06-29 11:50:26.000000000 +1000
@@ -96,6 +96,9 @@
   /** Per-mutator allocator into the large object space */
   protected LargeObjectLocal los = new LargeObjectLocal(Plan.loSpace);
 
+  /** Current dynamic allocation scope */
+  private int currentScope = Plan.DEFAULT_SCOPE;
+
   /****************************************************************************
    * 
    * Collection.
@@ -111,6 +114,14 @@
   public abstract void collectionPhase(int phaseId, boolean primary);
 
   /****************************************************************************
+   * Scoped allocation policy
+   */
+  public void setAllocationScope(int scope) {
+    currentScope = scope;
+  }
+
+
+  /****************************************************************************
    * 
    * Allocation.
    */
diff -r -w -u -N -I '[$]Id:' -I '[$]Revision:' -I '[$]Date:' --exclude='*.orig' --exclude=CVS --exclude='*.rej' --exclude='*~' base/MMTk/src/org/mmtk/plan/Plan.java MMTk/src/org/mmtk/plan/Plan.java
--- base/MMTk/src/org/mmtk/plan/Plan.java	2006-06-29 10:28:02.000000000 +1000
+++ MMTk/src/org/mmtk/plan/Plan.java	2006-06-29 11:50:26.000000000 +1000
@@ -122,6 +122,12 @@
   /** Support for time-limited GCs */
   protected static long timeCap;
 
+  /** 
+   * Support for scope-based allocation policy
+   */
+  public static final int DEFAULT_SCOPE = 1;
+  public static final int COMPILER_SCOPE = 2;
+  
   static {}
 
   /****************************************************************************
