Index: rvm/src/tools/bootImageRunner/syswrap.C
===================================================================
--- rvm/src/tools/bootImageRunner/syswrap.C	(revision 10849)
+++ rvm/src/tools/bootImageRunner/syswrap.C	(revision 10850)
@@ -214,7 +214,7 @@
 
     // Figure out how many seconds to wait
     double totalWaitTime;
-    if (timeout == 0)
+    if (timeout == NULL)
         totalWaitTime = VM_ThreadEventConstants_WAIT_INFINITE;
     else {
         totalWaitTime = ((double) timeout->tv_sec);
@@ -299,7 +299,7 @@
             FD_SET( ufds[i].fd, &exceptfds );
     }
 
-    ready = select(max_fd, &readfds, &writefds, &exceptfds, &tv);
+    ready = select(max_fd, &readfds, &writefds, &exceptfds, tv_ptr);
     
     for (unsigned i = 0; i < nfds; i++) {
 
